sql-plugin/src/main/scala/com/nvidia/spark/rapids/limit.scala

```diff
@@ -248,54 +248,72 @@ object GpuTopN {
       inputRows: GpuMetric,
       outputBatches: GpuMetric,
       outputRows: GpuMetric,
-      offset: Int): Iterator[ColumnarBatch] =
-    new Iterator[ColumnarBatch]() {
+      offset: Int): Iterator[SpillableColumnarBatch] =
+    new Iterator[SpillableColumnarBatch]() {
       override def hasNext: Boolean = iter.hasNext
 
       private[this] var pending: Option[SpillableColumnarBatch] = None
 
-      override def next(): ColumnarBatch = {
-        while (iter.hasNext) {
-          val input = iter.next()
-          withResource(new NvtxWithMetrics("TOP N", NvtxColor.ORANGE, opTime)) { _ =>
-            inputBatches += 1
-            inputRows += input.numRows()
-            lazy val totalSize = GpuColumnVector.getTotalDeviceMemoryUsed(input) +
-                pending.map(_.sizeInBytes).getOrElse(0L)
-
-            val runningResult = if (pending.isEmpty) {
-              withResource(input) { input =>
-                apply(limit, sorter, input, sortTime)
-              }
-            } else if (totalSize > Integer.MAX_VALUE) {
-              // The intermediate size is likely big enough we don't want to risk an overflow,
-              // so sort/slice before we concat and sort/slice again.
-              val tmp = withResource(input) { input =>
-                apply(limit, sorter, input, sortTime)
-              }
-              withResource(concatAndClose(pending.get, tmp, concatTime)) { concat =>
-                apply(limit, sorter, concat, sortTime)
-              }
-            } else {
-              // The intermediate size looks like we could never overflow the indexes so
-              // do it the more efficient way and concat first followed by the sort/slice
-              withResource(concatAndClose(pending.get, input, concatTime)) { concat =>
-                apply(limit, sorter, concat, sortTime)
+      override def next(): SpillableColumnarBatch = {
+        if (!hasNext) {
+          throw new NoSuchElementException()
+        }
+        closeOnExcept(pending) { _ =>
+          while (iter.hasNext) {
+            val inputScb = closeOnExcept(iter.next()) { cb =>
+              inputBatches += 1
+              inputRows += cb.numRows()
+              SpillableColumnarBatch(cb, SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
+            }
+            withResource(new NvtxWithMetrics("TOP N", NvtxColor.ORANGE, opTime)) { _ =>
+              withRetry(inputScb, splitSpillableInHalfByRows) { attempt =>
+                val inputCb = attempt.getColumnarBatch()
+                if (pending.isEmpty) {
+                  withResource(inputCb) { _ =>
+                    apply(limit, sorter, inputCb, sortTime)
```

Comment:
We are nesting retries here. I don't think that really works properly. Especially with the current state machine.

@abellina have you tried anything like this? Do we need checks in the state machine to prevent it?

Reply:
Hi, I think there is no nested retrying. `apply` here is still using the original `fullySortBatch` method, which has no retry support. 

@abellina also told me the nested retry blocks will not work as we expect. Then I was keeping in mind to avoid nested retry blocks. What's why I don't update this `fullySortBatch` method in place to add retry support, instead introduce a new API named `fullySortBatchAndCloseWithRetry`, and replace the call to `fullySortBatch` only in 'GpuSortEachBatchIterator'.



I explain this in the PR description. FYI. 
> GpuTopN is a little complicated. It is not a good idea to leverage this new API fullySortBatchAndCloseWithRetry, otherwise we will get a nested retry block. Instead we only need to retry the whole processing that returns a pending result for the next iteration. It also supports Split and Retry. Besides, retry is also added to the processing of applying the offset after getting the final result.